File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/VideoRendererFactoryController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.FlutterRtcVideoRenderer
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry

/**
 * Controller of creating new [FlutterRtcVideoRenderer]s.
 *
 * @property messenger Messenger used for creating new [MethodChannel]s.
 * @property textureRegistry Registry to create new textures with.
 */
class VideoRendererFactoryController(
    private val messenger: BinaryMessenger,
    private val textureRegistry: TextureRegistry
) : MethodChannel.MethodCallHandler {
  /** Channel listened for the [MethodCall]s. */
  private val chan = MethodChannel(messenger, ChannelNameGenerator.name("VideoRendererFactory", 0))

  init {
    chan.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "create" -> {
        val renderer = FlutterRtcVideoRenderer(textureRegistry)
        result.success(VideoRendererController(messenger, renderer).asFlutterResult())
      }
    }
  }
}

File: ios/Classes/controller/VideoRendererFactoryController.swift
import Flutter

/// Controller creating new `FlutterRtcVideoRenderer`s.
class VideoRendererFactoryController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Flutter texture registry for creating new Flutter textures.
  private var registry: FlutterTextureRegistry

  /// Initializes a new `VideoRendererFactoryController`.
  init(messenger: FlutterBinaryMessenger, registry: FlutterTextureRegistry) {
    let channelName = ChannelNameGenerator.name(
      name: "VideoRendererFactory",
      id: 0
    )
    self.messenger = messenger
    self.registry = registry
    self.channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: messenger
    )
    self.channel.setMethodCallHandler { call, result in
      self.onMethodCall(call: call, result: result)
    }
  }

  /// Handles all the supported Flutter method calls.
  ///
  /// Creates new `FlutterRtcVideoRenderer`s.
  func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "create":
      let renderer = FlutterRtcVideoRenderer(registry: self.registry)
      result(
        VideoRendererController(messenger: self.messenger, renderer: renderer)
          .asFlutterResult()
      )
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
